Make it possible to define generic visitors

**Is your feature request related to a problem? Please describe.**
There is a high possibility that visitors with the same "shape" will be reused a lot. This can partially be solved by #68 , but I think it'd still be nice to be able to generate generic visitors.

**Describe the solution you'd like**
There should be a possibility to reference to the defined generic parameters. The problem is that attributes can't reference the generic parameters declared inside the classes:
```cs
[SyntaxTreeVisitor(typeof(Node), ReturnType = typeof(TValue))] // error CS0416: 'TValue': an attribute argument cannot use type parameters
public abstract partial class MyVisitorBase<TValue> { /* ... */ }
```

**Describe alternatives you've considered**
A very simple solution could be to introduce special marker types to reference the first, second, third, ... (8 should be plenty) generic parameters, like so:
```cs
[SyntaxTreeVisitor(typeof(Node), ReturnType = typeof(GenericParam.P1))] // P1 will reference TValue
public abstract partial class MyVisitorBase<TValue> { /* ... */ }
```

Another simple solution is referencing them by name in a separate property on the attribute. The advantage here is that `nameof` works from the outside:
```cs
[SyntaxTreeVisitor(typeof(Node), GenericReturnType = nameof(TValue))]
public abstract partial class MyVisitorBase<TValue> { /* ... */ }
```
This latter case would only complicate in the fact that `ReturnType` and `GenericReturnType` have to be orthogonally specified.

--- a/Sources/Core/Yoakke.Parser.Generator/ParserSourceGenerator.cs
+++ b/Sources/Core/Yoakke.Parser.Generator/ParserSourceGenerator.cs
@@ -149,9 +149,10 @@ namespace Yoakke.Parser.Generator
             }
 
             var (prefix, suffix) = symbol.ContainingSymbol.DeclareInsideExternally();
+            var (genericTypes, genericConstraints) = symbol.GetGenericCrud();
             return $@"
 {prefix} 
-partial {symbol.GetTypeKindName()} {className} : {TypeNames.ParserBase}
+partial {symbol.GetTypeKindName()} {className}{genericTypes} : {TypeNames.ParserBase} {genericConstraints}
 {{
     public {className}({TypeNames.ILexer} lexer)
         : base(lexer)

--- a/Sources/Core/Yoakke.SourceGenerator.Common/GeneratorBase.cs
+++ b/Sources/Core/Yoakke.SourceGenerator.Common/GeneratorBase.cs
@@ -108,7 +108,7 @@ namespace Yoakke.SourceGenerator.Common
         /// </summary>
         /// <param name="fileName">The name of the file to add.</param>
         /// <param name="text">The contents of the file to add.</param>
-        protected void AddSource(string fileName, string text) => this.Context.AddSource(fileName, text);
+        protected void AddSource(string fileName, string text) => this.Context.AddSource(SanitizeFileName(fileName), text);
 
         /// <summary>
         /// Requires a library to be referenced by the user.
@@ -190,5 +190,9 @@ namespace Yoakke.SourceGenerator.Common
             }
             return true;
         }
+
+        private static string SanitizeFileName(string str) => str
+            .Replace("<", "_lt_")
+            .Replace(">", "_gt_");
     }
 }

--- a/Sources/Core/Yoakke.SourceGenerator.Common/RoslynExtensions/SymbolExtensions.cs
+++ b/Sources/Core/Yoakke.SourceGenerator.Common/RoslynExtensions/SymbolExtensions.cs
@@ -73,11 +73,13 @@ namespace Yoakke.SourceGenerator.Common.RoslynExtensions
             void DeclareInsideExternallyRec(ISymbol symbol)
             {
                 if (symbol is null) return;
-                if (symbol is ITypeSymbol type)
+                if (symbol is INamedTypeSymbol type)
                 {
                     if (!type.IsPartial()) throw new InvalidOperationException("Non-partial type nesting");
                     DeclareInsideExternallyRec(symbol.ContainingSymbol);
-                    prefixBuilder!.AppendLine($"partial {type.GetTypeKindName()} {type.Name} {{");
+
+                    var (genericTypes, genericConstraints) = type.GetGenericCrud();
+                    prefixBuilder!.AppendLine($"partial {type.GetTypeKindName()} {type.Name}{genericTypes} {genericConstraints} {{");
                     suffixBuilder!.AppendLine("}");
                     return;
                 }
@@ -92,6 +94,37 @@ namespace Yoakke.SourceGenerator.Common.RoslynExtensions
 
             DeclareInsideExternallyRec(symbol);
             return (prefixBuilder.ToString(), suffixBuilder.ToString());
+        }
+
+        /// <summary>
+        /// Retrieves the generic crud needed for a type definition.
+        /// Useful for partial types
+        /// </summary>
+        /// <param name="symbol">The type symbol to get the crud for.</param>
+        /// <returns>The pair of type parameter list and generic constraints as strings.</returns>
+        public static (string TypeParameters, string Constraints) GetGenericCrud(this INamedTypeSymbol symbol)
+        {
+            if (symbol.TypeParameters.Length == 0) return (string.Empty, string.Empty);
+
+            var typeParams = $"<{string.Join(", ", symbol.TypeParameters.Select(p => p.Name))}>";
+
+            var constraints = new StringBuilder();
+            foreach (var param in symbol.TypeParameters)
+            {
+                // First type constraints
+                var constraintList = param.ConstraintTypes.Select(t => t.ToDisplayString()).ToList();
+
+                // Then all other
+                if (param.HasConstructorConstraint) constraintList.Add("new()");
+                if (param.HasNotNullConstraint) constraintList.Add("notnull");
+                if (param.HasReferenceTypeConstraint) constraintList.Add("class");
+                if (param.HasUnmanagedTypeConstraint) constraintList.Add("unmanaged");
+                if (param.HasValueTypeConstraint) constraintList.Add("struct");
+
+                if (constraintList.Count > 0) constraints.AppendLine($"where {param.Name} : {string.Join(", ", constraintList)}");
+            }
+
+            return (typeParams, constraints.ToString());
         }
 
         /// <summary>

--- a/Sources/Core/Yoakke.SyntaxTree.Generator/Diagnostics.cs
+++ b/Sources/Core/Yoakke.SyntaxTree.Generator/Diagnostics.cs
@@ -1,0 +1,43 @@
+// Copyright (c) 2021 Yoakke.
+// Licensed under the Apache License, Version 2.0.
+// Source repository: https://github.com/LanguageDev/Yoakke
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.CodeAnalysis;
+
+namespace Yoakke.SyntaxTree.Generator
+{
+    /// <summary>
+    /// Diagnostics for the syntax tree generator.
+    /// </summary>
+    internal static class Diagnostics
+    {
+#pragma warning disable RS2008 // Enable analyzer release tracking
+
+        /// <summary>
+        /// Happens, when the visitor specifies both a return type and a generic return type.
+        /// </summary>
+        public static readonly DiagnosticDescriptor BothReturnTypeAndGenericReturnTypeSpecified = new(
+            id: "YKSYNTAXTREEGEN001",
+            title: "Cannot specify both a return type and a generic return type",
+            messageFormat: "Cannot specify both a return type and a generic return type for a visitor",
+            category: "Yoakke.SyntaxTree.Generator",
+            DiagnosticSeverity.Error,
+            isEnabledByDefault: true);
+
+        // <summary>
+        /// Happens, when the visitor specifies a generic return type that is not defined by the visitor.
+        /// </summary>
+        public static readonly DiagnosticDescriptor GenericTypeNotDefined = new(
+            id: "YKSYNTAXTREEGEN002",
+            title: "Generic type not defined",
+            messageFormat: "The class '{0}' does not define a generic parameter named '{1}'",
+            category: "Yoakke.SyntaxTree.Generator",
+            DiagnosticSeverity.Error,
+            isEnabledByDefault: true);
+
+#pragma warning restore RS2008 // Enable analyzer release tracking
+    }
+}

--- a/Sources/Core/Yoakke.SyntaxTree.Generator/SyntaxTreeSourceGenerator.cs
+++ b/Sources/Core/Yoakke.SyntaxTree.Generator/SyntaxTreeSourceGenerator.cs
@@ -43,6 +43,8 @@ namespace Yoakke.SyntaxTree.Generator
 
             public INamedTypeSymbol? ReturnType { get; set; }
 
+            public string? GenericReturnType { get; set; }
+
             public string? MethodName { get; set; }
         }
 
@@ -111,10 +113,16 @@ namespace Yoakke.SyntaxTree.Generator
                 var visitor = new Visitor(symbol!);
                 foreach (var attr in attrs)
                 {
+                    if (attr.ReturnType is not null && attr.GenericReturnType is not null)
+                    {
+                        this.Report(Diagnostics.BothReturnTypeAndGenericReturnTypeSpecified, syntax.GetLocation());
+                        continue;
+                    }
+
                     visitor.Overrides[attr.NodeType!] = new(attr.NodeType!)
                     {
                         MethodName = attr.MethodName,
-                        ReturnType = attr.ReturnType,
+                        ReturnType = (object?)attr.ReturnType ?? attr.GenericReturnType,
                     };
                 }
                 visitors.Add(visitor);
@@ -190,6 +198,18 @@ namespace Yoakke.SyntaxTree.Generator
                 .Select(m => (m.Name, m.Parameters[0].Type))
                 .ToHashSet(NameSymbolPairComparer.Instance);
 
+            ITypeSymbol GetGenericParam(string name)
+            {
+                foreach (var param in visitor.VisitorClass.TypeParameters)
+                {
+                    if (param.Name == name) return param;
+                }
+
+                // Not found, error
+                this.Report(Diagnostics.GenericTypeNotDefined, visitor.VisitorClass.Locations[0], visitor.VisitorClass.Name, name);
+                return voidType;
+            }
+
             void GenerateMembersVisitor(MetaNode currentNode)
             {
                 var enumerableInterface = this.LoadSymbol(TypeNames.IEnumerable);
@@ -226,7 +246,13 @@ namespace Yoakke.SyntaxTree.Generator
                 if (!currentNode.VisitorOverrides.TryGetValue(visitor, out var overrides)) return;
 
                 // Otherwise we need to generate code, as the visitor deals with this node
-                var returnType = overrides.ReturnType ?? voidType;
+                var returnType = overrides.ReturnType switch
+                {
+                    INamedTypeSymbol t => t,
+                    string genericParamName => GetGenericParam(genericParamName),
+                    null => voidType,
+                    _ => throw new InvalidOperationException(),
+                };
                 var returnTypeStr = returnType.ToDisplayString();
                 var methodName = overrides.MethodName ?? "Visit";
 

--- a/Sources/Core/Yoakke.SyntaxTree.Generator/VisitorOverride.cs
+++ b/Sources/Core/Yoakke.SyntaxTree.Generator/VisitorOverride.cs
@@ -26,8 +26,9 @@ namespace Yoakke.SyntaxTree.Generator
 
         /// <summary>
         /// Override for the method return type.
+        /// Can be either a <see cref="INamedTypeSymbol"/>, or a generic parameter name as a <see cref="string"/>.
         /// </summary>
-        public INamedTypeSymbol? ReturnType { get; set; }
+        public object? ReturnType { get; set; }
 
         /// <summary>
         /// Initializes a new instance of the <see cref="VisitorOverride"/> class.

--- a/Sources/Core/Yoakke.SyntaxTree/Attributes/SyntaxTreeVisitorAttribute.cs
+++ b/Sources/Core/Yoakke.SyntaxTree/Attributes/SyntaxTreeVisitorAttribute.cs
@@ -20,7 +20,12 @@ namespace Yoakke.SyntaxTree.Attributes
         /// <summary>
         /// The type to return from the visitor calls.
         /// </summary>
-        public Type ReturnType { get; set; }
+        public Type? ReturnType { get; set; }
+
+        /// <summary>
+        /// The generic type name to return from the visitor calls.
+        /// </summary>
+        public string? GenericReturnType { get; set; }
 
         /// <summary>
         /// The method name to use. The default is 'Visit'.
@@ -33,7 +38,7 @@ namespace Yoakke.SyntaxTree.Attributes
         /// <param name="nodeType">The type of the node the visitor visits.</param>
         /// <param name="returnType">The type to return from the visitor calls.</param>
         /// <param name="methodName">The method name to use.</param>
-        public SyntaxTreeVisitorAttribute(Type nodeType, Type returnType, string? methodName = null)
+        public SyntaxTreeVisitorAttribute(Type nodeType, Type? returnType, string? methodName = null)
         {
             this.NodeType = nodeType;
             this.ReturnType = returnType;
@@ -45,7 +50,7 @@ namespace Yoakke.SyntaxTree.Attributes
         /// </summary>
         /// <param name="nodeType">The type of the node the visitor visits.</param>
         public SyntaxTreeVisitorAttribute(Type nodeType)
-            : this(nodeType, typeof(void))
+            : this(nodeType, null)
         {
         }
     }

--- a/Sources/Tests/Yoakke.SyntaxTree.Tests/GenericVisitorTests.cs
+++ b/Sources/Tests/Yoakke.SyntaxTree.Tests/GenericVisitorTests.cs
@@ -37,7 +37,7 @@ namespace Yoakke.SyntaxTree.Tests
             protected override int Visit(Ast.Bar bar) => 2;
         }
 
-        private class StringVisitor : GenericVisitorBase<int>
+        private class StringVisitor : GenericVisitorBase<string>
         {
             public string GetString(Ast ast) => this.Visit(ast);
 

--- a/Sources/Tests/Yoakke.SyntaxTree.Tests/GenericVisitorTests.cs
+++ b/Sources/Tests/Yoakke.SyntaxTree.Tests/GenericVisitorTests.cs
@@ -1,0 +1,65 @@
+// Copyright (c) 2021 Yoakke.
+// Licensed under the Apache License, Version 2.0.
+// Source repository: https://github.com/LanguageDev/Yoakke
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Yoakke.SyntaxTree.Attributes;
+
+namespace Yoakke.SyntaxTree.Tests
+{
+    [TestClass]
+    public partial class GenericVisitorTests
+    {
+        [SyntaxTree]
+        internal abstract partial record Ast
+        {
+            public partial record Foo : Ast;
+
+            public partial record Bar : Ast;
+        }
+
+        [SyntaxTreeVisitor(typeof(Ast), GenericReturnType = nameof(TReturn))]
+        internal abstract partial class GenericVisitorBase<TReturn>
+        {
+        }
+
+        private class IntVisitor : GenericVisitorBase<int>
+        {
+            public int GetInt(Ast ast) => this.Visit(ast);
+
+            protected override int Visit(Ast.Foo foo) => 1;
+
+            protected override int Visit(Ast.Bar bar) => 2;
+        }
+
+        private class StringVisitor : GenericVisitorBase<int>
+        {
+            public string GetString(Ast ast) => this.Visit(ast);
+
+            protected override string Visit(Ast.Foo foo) => "foo";
+
+            protected override string Visit(Ast.Bar bar) => "bar";
+        }
+
+        [TestMethod]
+        public void IntVisitorTests()
+        {
+            var visitor = new IntVisitor();
+            Assert.AreEqual(1, visitor.GetInt(new Ast.Foo()));
+            Assert.AreEqual(2, visitor.GetInt(new Ast.Bar()));
+        }
+
+        [TestMethod]
+        public void StringVisitorTests()
+        {
+            var visitor = new StringVisitor();
+            Assert.AreEqual("foo", visitor.GetString(new Ast.Foo()));
+            Assert.AreEqual("bar", visitor.GetString(new Ast.Bar()));
+        }
+    }
+}

--- a/Sources/Tests/Yoakke.SyntaxTree.Tests/VisitorTests.cs
+++ b/Sources/Tests/Yoakke.SyntaxTree.Tests/VisitorTests.cs
@@ -27,7 +27,9 @@ namespace Yoakke.SyntaxTree.Tests
         }
 
         [SyntaxTreeVisitor(typeof(Ast))]
-        internal abstract partial class VoidVisitorBase { }
+        internal abstract partial class VoidVisitorBase
+        {
+        }
 
         internal class MyVoidVisitor : VoidVisitorBase
         {
@@ -51,7 +53,9 @@ namespace Yoakke.SyntaxTree.Tests
         }
 
         [SyntaxTreeVisitor(typeof(Ast), typeof(string))]
-        internal abstract partial class StrVisitorBase { }
+        internal abstract partial class StrVisitorBase
+        {
+        }
 
         internal class MyStrVisitor : StrVisitorBase
         {

